atc/lidar/interval_runner.go

```diff
@@ -56,30 +60,61 @@ func (r *intervalRunner) Run(signals <-chan os.Signal, ready chan<- struct{}) er
 
 	close(ready)
 
-	notifier, err := r.notifications.Listen(r.channel)
+	notifier, err := r.notifications.Listen(r.componentName)
 	if err != nil {
 		return err
 	}
 
-	defer r.notifications.Unlisten(r.channel, notifier)
+	defer r.notifications.Unlisten(r.componentName, notifier)
 
 	ticker := r.clock.NewTicker(r.interval)
 	ctx, cancel := context.WithCancel(context.Background())
 
-	if err := r.runner.Run(ctx); err != nil {
-		r.logger.Error("failed-to-run", err)
-	}
-
 	for {
 		select {
 		case <-ticker.C():
+			component, _, err := r.componentFactory.Find(r.componentName)
```

Comment:
Hey,

If I understood this right, this is inherintly racy in a way that would result
in the same behavior as before :thinking:

```
time 		atc1			atc2
0		find
1					find
2		oh not paused
3					oh not paused
4		interval elapsed
5					interval elapsed
6		runs
7					runs

```

in a way that we get two components running for the same interval.

Is that right? Or, am I missing another synchronization that might be happening
implicity somewhere else? 

Thanks!

Reply:
There is a "task" (aka component) lock that already being using in `atc/lockrunner/runner.go` for gc, build-reaper and syslog-drainer. I think we should do the same for build-tracker and lidar scanner/checker as well. Will update the code accordingly.

Reply:
@xtremerui @cirocosta 

I don't think we need this extra locking because the underlying component does the locking itself: https://github.com/concourse/concourse/blob/master/atc/lidar/scanner.go#L43

Reply:
I can see the argument for acquiring the lock first, but I'm really not worried about a race condition here. Plus there's no way to avoid that race for things like the build tracker, and resource checker which grab locks at a lower level for each individual build/check.

Reply:
@pivotal-jwinters there is a lock here in lock runner 
https://github.com/concourse/concourse/blob/517bc6f3343bfbafa6c52b884f5b4bfced726d4c/atc/lockrunner/runner.go#L59

I'd say we could make them consistent at least? Could we remove that lock in this case?

Reply:
Yeah the lock runner has its own lock because the underlying component doesn't do its own locking. I agree we should make this consistent. I like the idea of having all the locking in the runner, but that won't work for the build tracker and resource checker because they do locking at a lower level. Maybe each component should be responsible for its own locking and we get rid of the lock runner?

Reply:
> I don't think we need this extra locking because the underlying component does the locking itself: https://github.com/concourse/concourse/blob/master/atc/lidar/scanner.go#L43

isn't that locking just the `run` though?

 I'm just afraid that locking just at that late stage would make us get to the situation that we had before where one would run immediately  after the other given that they could overlap the checking the interval + pause as those operations have no mutual exclusion (as shown in the ascii diagram) 🤔

Reply:
@cirocosta it's locks all the way down.

> I guess we could lock on the entire component and then on each sub task, but that seems unnecessary.

After thinking about this more, I'm kinda backtracking on this. Maybe it is a good idea to lock at multiple levels. It's kinda like you said, lock the `run` then sub tasks can do their own locking if they need to.

Reply:
@cirocosta we now do the locking up front. This adds another layer of locking around certain components, but that seems like an ok tradeoff.